SSZ-QL: recursive hash tree root computation - basic types, fixed size fields, containers & lists

fernantho opened the pull request:
**What type of PR is this?**
Feature

**What does this PR do? Why is it needed?**
This PR initializes the **hash tree root** computation for SSZ-QL. The focus of this PR is to compute the hash tree root for:
- basic types
- containers
- lists
- vectors

This PR does not handle:
- unions
- progressive Lists

**Which issues(s) does this PR fix?**

Part of #15598. 

**Other notes for review**
- Scaffolds a recursive function to compute hash tree roots for arbitrary objects, then iterates on specialized builders for lists/vectors/containers. 
- Implements HTR algorithms for basic types, containers, and lists; later decomposes list/vector handling per subtype (basic, composite, bitvector/bitlist).
- Added hash tree root algorithms for basic types, containers and lists.
- Empty bitlists are to be added once fastssz serializes them correctly.

**Acknowledgements**

- [x] I have read [CONTRIBUTING.md](https://github.com/prysmaticlabs/prysm/blob/develop/CONTRIBUTING.md).
- [x] I have included a uniquely named [changelog fragment file](https://github.com/prysmaticlabs/prysm/blob/develop/CONTRIBUTING.md#maintaining-changelogmd).
- [x] I have added a description to this PR with sufficient context for reviewers to understand this PR.

fernantho committed:
exported types and fields needed for the hash tree root

fernantho committed:
added hash tree root core recursive algo scaffold. added hash tree root for basic types

fernantho committed:
added tests for hash tree root for basic types

fernantho committed:
run gazelle

fernantho committed:
added changelog

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/OffchainLabs/prysm?pullRequest=15669) <br/>All committers have signed the CLA.

fernantho committed:
added hash tree root for containers composed of basic type fields

fernantho committed:
added tests for containers

fernantho committed:
added hash tree root for basic-type vectors

fernantho committed:
removed bytesPerChunk constant

fernantho committed:
removed outdated instruction

fernantho committed:
removed duplicated offset exported func

fernantho committed:
refactored computeContainerHashTreeRoot - improved sanity check and took into account variable field length

fernantho committed:
more cleaning before refactoring computeListHashTreeRoot

fernantho committed:
updated BUILD.bazel

fernantho committed:
fixed lists, containers and vectors hash tree roots

fernantho committed:
Merge branch 'develop' into feat/ssz-query-initialize-recursive-hash-tree-root

fernantho committed:
refactoring hash tree root computation in favor of current prysm hasher implementation

fernantho committed:
cleaning idt for verbose indentation. to added at the end

fernantho committed:
updated call to HashTreeRoot(...) in test + completed flow for lists/vectors of composite types

fernantho committed:
minor test renaming

fernantho committed:
extracted switch/case logic into dedicated buildRoot* functions

fernantho committed:
fixed issue buildRootFromList wrong offset + slice computation

fernantho committed:
removed previous developments

fernantho committed:
dirty commit pre merge develop

fernantho committed:
merged develop into current branch

fernantho committed:
refactored bitlist computation. added custom type to test it detached from any other type.

fernantho committed:
added BitvectorContainer custom type. Fixed bitvector hash tree root computation. Added test

fernantho committed:
removed previously hardcoded value for testing

fernantho committed:
added minor fix for empty bitlists

fernantho committed:
discarded ProgressiveList processing

fernantho committed:
added new custom type - BasicTypeList

fernantho committed:
added bitlist test suite + fixed basic type list computation

fernantho committed:
working version for all hash tree root routines - need a heavy refactor + increase test coverage

fernantho committed:
added documentation + minor refactor

fernantho committed:
refactorization to improve quality

- checked every error & improved handling by increasing context.
- removed TODO in buildRootFromBasicType
- caught a wrong ProgressiveList condition
- added getFieldNames helper function

fernantho committed:
- decomposed list and vector functions per subtype: basic, composite or bitvector/bitlist
- reordered and cleaned tests

fernantho committed:
removed old patch for empty bitlists

fernantho committed:
merged develop into current branch
